src/lastcomm/lastcomm.1
.\" Copyright (c) 1980, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	From: @(#)lastcomm.1	8.1 (Berkeley) 6/6/93
.\"	$Id: lastcomm.1,v 1.3 1996/09/18 16:48:54 wollman Exp $
.\"
.Dd September 18, 1996
.TH LASTCOMM 1
.Os BSD 3
.SH NAME
lastcomm
\- show last commands executed in reverse order
.SH SYNOPSIS
.B lastcomm
[\fB-EScestu\fR]
[\fB-f\fI file\fR]
[\fIcommand ...\fR]
[\fIuser ...\fR]
[\fIterminal ...\fR]
.SH DESCRIPTION
.B Lastcomm
gives information on previously executed commands.
With no arguments,
.B lastcomm
prints information about all the commands recorded
during the current accounting file's lifetime.
.PP
The following options are available:
.IP \fB-E\fR
Print the time the process exited.
.IP \fB-S\fR
Print the time the process started.
.IP \fB-c\fR
Print the amount of cpu time used by the process.
.IP \fB-e\fR
Print the amount of elapsed time used by the process.
.IP \fB-s\fR
Print the amount of system time used by the process.
.IP \fB-t\fR
Produce output in \fBtime machine\fR format: a portable, self-documenting
file format (see tm-format(5)) that contains all accounting record fields.
.RS
.IP \fBcommand\fR
The command name.
.IP \fBflags\fR
Flags accumulated by the accounting system.
.IP \fBuid\fR
The numerical user ID.
.IP \fBgid\fR
The numerical group ID.
.IP \fBtty\fR
The controlling terminal, given as \fIminor, major\fR device numbers
or given as ``__'' for processes without controlling terminal.
.IP \fBuser_time\fR
User-mode CPU time in seconds.
.IP \fBsystem_time\fR
Kernel-mode CPU time in seconds.
.IP \fBelapsed_time\fR
Elapsed time in seconds.
.IP \fBmemory_usage\fR
Amount of memory used (FreeBSD, BSD/OS, SunOS 4 and Solaris 2).
.IP \fBcharacter_io\fR
Character device I/O (SunOS 4 and Solaris 2).
.IP \fBblock_io\fR
Block device I/O (FreeBSD, BSD/OS, SunOS 4 and Solaris 2).
.IP \fBmajor_page_faults\fR
Page fault statistics (Linux).
.IP \fBminor_page_faults\fR
Page fault statistics (Linux).
.IP \fBexit_status\fR
Process exit status (Linux).
.RE
.IP \fB-u\fR
Print the amount of user time used by the process.
.IP "\fB-f\fR \fIfile\fR"
Read from
.I file
rather than the default \fB/var/account/acct\fR.
.PP
If no options are specified, \fB-cS\fR is assumed.
If \fBlastcomm\fR is invoked with arguments, only accounting entries with a
matching \fIcommand\fR name, \fIuser\fR name, or \fIterminal\fR name
are printed.
For example:
.PP
.DS
lastcomm a.out root ttyd0
.DE
.PP
would produce a listing of all the
executions of commands named
\fIa.out\fR
by user
\fIroot\fR
on the terminal
\fIttyd0\fR.
.PP
For each process entry, the following are printed.
.IP \(bu
The name of the user who ran the process.
.IP \(bu
Flags, as accumulated by the accounting facilities in the system.
.IP \(bu
The command name under which the process was called.
.IP \(bu
The amount of 
CPU
(\fB-c\fR),
wall
(\fB-e\fR),
system
(\fB-s\fR),
or user
(\fB-u\fR)
time used by the process (in seconds).
.IP \(bu
The time the process started
(\fB-S\fR)
or exited
(\fB-E\fR).
.PP
The flags are encoded as follows: ``S'' indicates the command was
executed by the super-user, ``F'' indicates the command ran after
a fork, but without a following
exec ,
.\" ``C'' indicates the command was run in PDP-11 compatibility mode
.\" (VAX only),
``D'' indicates the command terminated with the generation of a
.Pa core
file, and ``X'' indicates the command was terminated with a signal.
.SH FILES
/var/account/acct, Default accounting file.
.SH SEE ALSO
last(1),
sigvec(2),
acct(5),
core(5)
tm-format(5)
.SH HISTORY
The
.B lastcomm
command appeared in 3.0BSD.


CHANGES
Sat Aug 29 2009

o	grave-robber did not record the initial directory in the body file.

o	Bit-rot: FreeBSD 7 support, including ``struct acct'' changes.

o	Bit-rot: lastcomm format strings.

Sun Jan 11 20:01:27 EST 2009

o	Bit-rot: OpenBSD 4.x support (still TODO: file system tools).

Thu Jun  7 10:54:44 EDT 2007

o	Bit-rot: FreeBSD 6.x support.

Mon Apr  9 09:36:31 EDT 2007

o	Bit-rot: the 64-bit workaround for LINUX lseek() is no
	longer needed, and it didn't compile on some systems.

Thu Jan 11 16:10:29 EST 2007

o	Bugfix: the timeout command always reported zero exit status
	after voluntary child process exit.

Mon Jun 26 18:44:17 EDT 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Fri Mar 17 09:25:57 EST 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Tue Jan  6 12:25:32 EST 2004

o       Workaround: mactime treated "," as a special character in
	file names, now it's 0xff.

o	Support for ext[23]fs inode file sizes > 32bits.

o	Bugfix: ils file sizes larger than unsigned long.

Tue Oct 14 18:23:49 EDT 2003

o       pcat now fully supports FreeBSD systems that do not have
	/proc mounted (it's no longer mounted by default with
	FreeBSD 5). pcat tries to use /proc if it can, and uses
	ptrace() and gropes kernel memory if it has to.  Operation
	without /proc requires super-user privileges.

Thu Oct  2 09:14:00 EDT 2003

o       Completed support for UFS1 and UFS2 in FreeBSD 5.x. However,
	the pcat command still requires that the /proc file system
	is mounted. This will be fixed later.

Sat Aug 30 19:21:29 EDT 2003

o	Preliminary FreeBSD 5.0 port. UFS2 support is still to be done.

Fri Aug  1 10:07:39 EDT 2003

o	RedHat 9.0 fix by Florin Andrei for the file command.

Sat Oct  5 13:47:29 EDT 2002

o	Restored tctutils compatibility, which broke with 20020916.

Mon Sep 16 14:39:09 EDT 2002

o       The lowest inode examined is now the lowest possible inode
	number instead of ROOTINUM.

Fri Mar 22 14:41:32 EST 2002

o	Added support for OpenBSD 3.0.

Wed Dec 12 13:21:00 EST 2001

o	tm_print() and tm_string() did unnecessary %HEX expansion
	of $ and ^ characters (Philippe Bourgeois).

o       mactime mis-sorted dates before/after early September 2001
	(Philippe Bourgeois, Brian Carrier).

Thu Oct  4 15:48:42 EDT 2001

o       Linux can handle files > 2Gbyte, but only if we specify
	extra compiler flags.

o	Solaris 8/64bit needs an extra include file in pcat.c.

Sun Sep  9 17:23:13 EDT 2001

o       Discovered to my horror that unrm on ext2fs can have an
	off-by-one error when testing a bit flag in the data block
	allocation bitmaps - and this problem is dependent on the
	Linux version being used.

o       Made the timeout command a little more verbose so it is
	less of a mystery when the child process is killed.

Thu Sep  6 13:51:00 EDT 2001

o	New grave-robber -I option for better control of icat/pcat.

Mon May 21 10:11:19 EDT 2001

o       findkey command-line error check was too conservative,
	and disallowed threshold of 256 different byte values
	(Ovanes Manucharyan).

Mon May 14 13:11:54 EDT 2001

o	Patch 06 broke lib/pass.cache.pl, which broke mactime.

o	Patch 06 did not update the patchlevel file.

Fri Mar 16 14:25:11 EST 2001

o	Fixed rpm root directory when running on a corpse.
	Marco Walter (SUN).

o	Fixed file(1) output parsing problem with multiple :
	characters.  Marco Walter (SUN)

Thu Feb 22 11:16:40 EST 2001

o	Patch by Brian Carrier (Purdue) so that fs_read_random()
	properly maintains file offset information.

Thu Feb  8 10:40:02 EST 2001

o	grave-robber dies silently when no group file is found.
	This bites when examining a partial machine corpse
	(bin/grave-robber). Klaus-Peter Kossakowski.

o	The "do first looks" code did not relocate to $TCT_HOME
	(lib/tree.pl). Klaus-Peter Kossakowski.

Tue Jan 23 18:44:36 EST 2001

o       The grave-robber did not correctly handle multi-line df
	output records (ps_spy.pl, suck_free_inodes.pl).

o       The grave-robber mounted corpse inode scan looked only at
	the corpse's "root" file system (suck_free_inodes.pl).

Mon Jan 22 11:23:32 EST 2001

o       grave-robber enabled the default -lPO flags when -c was
	specified, causing the program to gripe and exit.
	
Tue Sep 26 10:55:24 EDT 2000

o       Fixed a bug in fs_read_block() that caused icat etc. to
	remember the wrong file seek position, so that icat etc.
	would call lseek() too often. The other possibility, not
	calling lseek() when it should be called, did not happen
	due to an implementation artefact of the TCT utilities.
	Reported by Brian Carrier, Purdue University.

	Updated help-recovering-file document by Dan.

Sat Aug 12 10:51:16 EDT 2000

o       Added -V option to direct icat/ils/unrm for verbose output
	to stdout.  The -v option now sends verbose logging to stderr.

Tue Aug  8 11:46:19 PDT 2000

o	created "help-recovering-file", added note to README.FIRST

Sat Aug  5 11:34:26 PDT 2000

o	mactime time range repaired, fixed docs, -l option slightly
	fixed... perhaps deprecation in the future?

Fri Aug  4 20:06:15 EDT 2000

o	ils and ils2mac now include device names for more
	readable mactime reports of dead inodes.

o	mactime could not run with current working directory
	different from $TCT_HOME.

o	mactime -d could destroy an existing $DATA/body file.

o	mactime -d turned on verbosity, instead of letting -v do that.

o	mactime -s (suid) flag used an obsolete method of getting data.

o	ils2mac broke because perl split() can't see the difference
	between empty last fields and missing last fields.

o	mactime mistakenly skipped files with all-zero time stamps.

o	ils2mac now emits ls -l like info for mactime compatibility.

Tue Aug  1 13:50:20 PDT 2000

o	Prepend our directory list to @INC rather than clobbering it.

Mon Jul 31 17:37:35 PDT 2000

o	removed mention of sans.org from help-when-broken-into


src/lastcomm/lastcomm.c
/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *  $Id: lastcomm.c,v 1.4 1996/06/30 11:58:20 wosch Exp $
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1980, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)lastcomm.c	8.1 (Berkeley) 6/6/93";
#endif /* not lint */

#include "sys_defs.h"

#ifdef USE_SYSMACROS_H
#include <sys/sysmacros.h>
#endif

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/acct.h>

#include <time.h>
#include <ctype.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <struct.h>
#include <unistd.h>
#include <utmp.h>
/*#include "pathnames.h"*/

#ifdef HAVE_ACCTV2
#define ACCT_STRUCT	acctv2
#define ACCT_FLAG	ac_flagx
#define expand(x)	(x)
#define AHZ		1000000
#else
#define ACCT_STRUCT	acct
#define ACCT_FLAG	ac_flag
time_t	 expand(u_int);
#endif
char	*flagbits(int);
char	*getdev(dev_t);
int	 requested(char *[], struct ACCT_STRUCT *);
void	 usage(void);
char	*user_from_uid();

char	*print_uid(uid_t);
#define user_from_uid(u,f) print_uid(u)
#define err(e, fmt, s) { perror(s); exit(e); }

#define AC_UTIME 1 /* user */
#define AC_STIME 2 /* system */
#define AC_ETIME 4 /* elapsed */
#define AC_CTIME 8 /* user + system time, default */

#define AC_BTIME 16 /* starting time */
#define AC_FTIME 32 /* exit time (starting time + elapsed time )*/

#define AC_HZ ((double)AHZ)

#ifndef PRINT_NAMESIZE
#define PRINT_NAMESIZE	UT_NAMESIZE
#define PRINT_LINESIZE	UT_LINESIZE
#endif

int
main(argc, argv)
	int argc;
	char *argv[];
{
	register char *p;
	struct ACCT_STRUCT ab;
	struct stat sb;
	FILE *fp;
	off_t size;
	time_t t;
	int ch;
	char *acctfile;
	int time = 0;
	int time_machine = 0;

	acctfile = _PATH_ACCT;
	while ((ch = getopt(argc, argv, "f:tusecSE")) != EOF)
		switch((char)ch) {
		case 'f':
			acctfile = optarg;
			break;

		case 't':
			time_machine = 1;
			break;
		case 'u': 
			time |= AC_UTIME; /* user time */
			break;
		case 's':
			time |= AC_STIME; /* system time */
			break;
		case 'e':
			time |= AC_ETIME; /* elapsed time */
			break;
		case 'c':
			time |= AC_CTIME; /* user + system time */
			break;

		case 'S':
			time |= AC_BTIME; /* starting time */
			break;
		case 'E':
			/* exit time (starting time + elapsed time )*/
			time |= AC_FTIME; 
			break;

		case '?':
		default:
			usage();
		}

	/* default user + system time and starting time */
	if (!time) {
	    time = AC_CTIME | AC_BTIME;
	}

	argc -= optind;
	argv += optind;

	/* Open the file. */
	if ((fp = fopen(acctfile, "r")) == NULL || fstat(fileno(fp), &sb))
		err(1, "%s", acctfile);

	/*
	 * Round off to integral number of accounting records, probably
	 * not necessary, but it doesn't hurt.
	 */
	size = sb.st_size - sb.st_size % sizeof(struct ACCT_STRUCT);

	/* Print the optional time machine header. */
	if (time_machine) {
		char myhostname[BUFSIZ];
		time_t time();
		time_t now;
		if (gethostname(myhostname, sizeof(myhostname) - 1) < 0)
			err(1, "%s", "gethostname");
		time(&now);
		printf("class|host|start_time\n%s|%s|%lu\n",
			"lastcomm", myhostname, (unsigned long) now);
		printf("command|flags|uid|gid|tty");
		printf("|user_time|system_time|start_time|elapsed_time");
#if defined(HAVE_MEMORY_USAGE) || defined(HAVE_COMP_MEMORY_USAGE)
		printf("|memory_usage");
#endif
#ifdef HAVE_COMP_CHAR_IO_COUNT
		printf("|character_io");
#endif
#ifdef HAVE_COMP_BLOCK_IO_COUNT
		printf("|block_io");
#endif
#ifdef HAVE_COMP_BLOCK_RW_COUNT
		printf("|block_io");
#endif
#if defined(HAVE_MAJOR_PFLTS) || defined(HAVE_COMP_MAJOR_PFLTS)
		printf("|major_page_faults|minor_page_faults");
#endif
#ifdef HAVE_EXIT_STATUS
		printf("|exit_status");
#endif
#ifdef HAVE_COMP_SWAP_USAGE
		printf("|swaps");
#endif
		printf("\n");
	}

	/* Check if any records to display. */
	if (size < sizeof(struct ACCT_STRUCT))
		exit(0);

	/*
	 * Seek to before the last entry in the file; use lseek(2) in case
	 * the file is bigger than a "long".
	 */
	size -= sizeof(struct ACCT_STRUCT);
	if (lseek(fileno(fp), size, SEEK_SET) == -1)
		err(1, "%s", acctfile);

	for (;;) {
		if (size < 0)
			break;

		if (fread(&ab, sizeof(struct ACCT_STRUCT), 1, fp) != 1)
			err(1, "%s", acctfile);

		size -= sizeof(struct ACCT_STRUCT);

		if (size >= 0)
		if (fseek(fp, 2 * -(long)sizeof(struct ACCT_STRUCT), SEEK_CUR) == -1)
			err(1, "%s", acctfile);

		if (ab.ac_comm[0] == '\0') {
			ab.ac_comm[0] = '?';
			ab.ac_comm[1] = '\0';
		} 
		if (*argv && !requested(argv, &ab))
			continue;

		/* Optionally produce time machine format output */
#define UNSIG(x) ((unsigned char *)(x))
		if (time_machine) {
			for (p = ab.ac_comm; *p != 0
				&& p < ab.ac_comm + fldsiz(ACCT_STRUCT, ac_comm); p++)
				if (isascii(*UNSIG(p)) && isprint(*UNSIG(p))
				&& *UNSIG(p) != '|' && *UNSIG(p) != '%')
					putchar(*UNSIG(p));
				else
					printf("%%%02X", *UNSIG(p));
			printf("|%s|%lu|%lu",
				flagbits(ab.ACCT_FLAG),
				(unsigned long) ab.ac_uid,
				(unsigned long) ab.ac_gid);
			printf("|%s", getdev(ab.ac_tty));

			printf("|%.3f|%.3f|%ld|%.3f",
				expand(ab.ac_utime) / AC_HZ,
				expand(ab.ac_stime) / AC_HZ,
				(long) ab.ac_btime,
				expand(ab.ac_etime) / AC_HZ);

#ifdef HAVE_MEMORY_USAGE
			printf("|%lu", (unsigned long) ab.ac_mem);
#endif
#ifdef HAVE_COMP_MEMORY_USAGE
			printf("|%lu", (unsigned long) expand(ab.ac_mem));
#endif
#ifdef HAVE_COMP_CHAR_IO_COUNT
			printf("|%lu", (unsigned long) expand(ab.ac_io));
#endif
#ifdef HAVE_COMP_BLOCK_IO_COUNT
			printf("|%lu", (unsigned long) expand(ab.ac_io));
#endif
#ifdef HAVE_COMP_BLOCK_RW_COUNT
			printf("|%lu", (unsigned long) expand(ab.ac_rw));
#endif
#ifdef HAVE_MAJOR_PFLTS
			printf("|%lu|%lu",
				(unsigned long) ab.ac_majflt,
				(unsigned long) ab.ac_minflt);
#endif
#ifdef HAVE_COMP_MAJOR_PFLTS
			printf("|%lu|%lu",
				(unsigned long) expand(ab.ac_majflt),
				(unsigned long) expand(ab.ac_minflt));
#endif
#ifdef HAVE_EXIT_STATUS
			printf("|%ld", (long) ab.ac_exitcode);
#endif
#ifdef HAVE_COMP_SWAP_USAGE
			printf("|%lu", (unsigned long) expand(ab.ac_swaps));
#endif
			printf("\n");
			continue;
		}

		/* Produce the default output format. */
		for (p = &ab.ac_comm[0];
		    p < &ab.ac_comm[fldsiz(ACCT_STRUCT, ac_comm)] && *p; ++p)
			if (!isprint(*p))
				*p = '?';
		(void)printf("%-*.*s %-7s %-*s %-*s ",
			     fldsiz(ACCT_STRUCT, ac_comm),
			     fldsiz(ACCT_STRUCT, ac_comm),
			     ab.ac_comm,
			     flagbits(ab.ACCT_FLAG),
			     PRINT_NAMESIZE, user_from_uid(ab.ac_uid, 0),
			     PRINT_LINESIZE, getdev(ab.ac_tty));
		
		
		/* user + system time */
		if (time & AC_CTIME) {
			(void)printf("%6.3f secs ", 
				     (expand(ab.ac_utime) + 
				      expand(ab.ac_stime))/AC_HZ);
		}
		
		/* usr time */
		if (time & AC_UTIME) {
			(void)printf("%6.3f us ", expand(ab.ac_utime)/AC_HZ);
		}
		
		/* system time */
		if (time & AC_STIME) {
			(void)printf("%6.3f sy ", expand(ab.ac_stime)/AC_HZ);
		}
		
		/* elapsed time */
		if (time & AC_ETIME) {
			(void)printf("%8.3f es ", expand(ab.ac_etime)/AC_HZ);
		}
		
		/* starting time */
		if (time & AC_BTIME) {
			(void)printf("%.16s ", ctime((time_t *)&ab.ac_btime));
		}
		
		/* exit time (starting time + elapsed time )*/
		if (time & AC_FTIME) {
			t = ab.ac_btime;
			t += (time_t)(expand(ab.ac_etime)/AC_HZ);
			(void)printf("%.16s ", 
				     ctime(&t));
		}
		printf("\n");
 	}
 	exit(0);
}

#ifndef HAVE_ACCTV2

time_t
expand(t)
	u_int t;
{
	register time_t nt;

	nt = t & 017777;
	t >>= 13;
	while (t) {
		t--;
		nt <<= 3;
	}
	return (nt);
}

#endif

char *
flagbits(f)
	register int f;
{
	static char flags[20] = "-";
	char *p;

#define	BIT(flag, ch)	if (f & flag) *p++ = ch

	p = flags + 1;
	BIT(ASU, 'S');
	BIT(AFORK, 'F');
#ifdef ACOMPAT
	BIT(ACOMPAT, 'C');
#endif
#ifdef ACORE
	BIT(ACORE, 'D');
#endif
#ifdef AXSIG
	BIT(AXSIG, 'X');
#endif
	*p = '\0';
	return (flags);
}

int
requested(argv, acp)
	register char *argv[];
	register struct ACCT_STRUCT *acp;
{
	register char *p;

	do {
		p = user_from_uid(acp->ac_uid, 0);
		if (!strcmp(p, *argv))
			return (1);
		if ((p = getdev(acp->ac_tty)) && !strcmp(p, *argv))
			return (1);
		if (!strncmp(acp->ac_comm, *argv, fldsiz(ACCT_STRUCT, ac_comm)))
			return (1);
	} while (*++argv);
	return (0);
}

char *
getdev(dev)
	dev_t dev;
{
	static dev_t lastdev = (dev_t)-1;
	static char lastname[BUFSIZ];
	if (dev == NODEV)			/* Special case. */
		return ("__");
	if (dev == lastdev)			/* One-element cache. */
		return (lastname);
	lastdev = dev;
	sprintf(lastname, "%d,%d", (int) major(dev), (int) minor(dev));
	return (lastname);
}

void
usage()
{
	(void)fprintf(stderr,
	    "lastcomm [-EScestu] [ -f file ] [command ...] [user ...] [tty ...]\n");
	exit(1);
}

char *
print_uid(uid_t uid)
{
	static char buf[BUFSIZ];

	sprintf(buf, "%lu", (unsigned long) uid);
	return (buf);
}
